### PR TITLE
Add TypeScript definition

### PR DESCRIPTION
**Pull request by BendingBender**

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+	/**
+	 * Flags converted camelCase including aliases.
+	 */
+	unnormalizedFlags: {[name: string]: unknown};
+
+	/**
+	 * The `package.json` object.
+	 */
+	pkg: {[key: string]: unknown};
```
Comment (sindresorhus) [inline index.d.ts]: Would be really cool if we could properly type the package.json.

https://docs.npmjs.com/files/package.json
http://json.schemastore.org/package

You could submit the definition to https://github.com/sindresorhus/type-fest so we could reuse it in other packages too.

What do you think?

---

We could also expose a type that extends the spec compliant one with a realistic definition that also supports keys like `types` and `eslintConfig`.

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+	 * 	}
+	 * }
+	 */
+	readonly flags?: MinimistOptions;
+
+	/**
+	 * Description to show above the help text. Default: The package.json `"description"` property.
+	 *
+	 * Set it to `false` to disable it altogether.
+	 */
```
[the remark this answers is not in the document]
Comment (bfred-it) [reply index.d.ts]: I think they are the JSDoc style. Would this break VSCode's ability to display the description inline, for example?

> hard to modify the comments

That just depends on the editor and its plugins. I had no problem with Sublime Text (but I think I'm missing some plugin in VSCode)

[the remark this answers is not in the document]
Comment (BendingBender) [reply index.d.ts]: This, however, disables markdown syntax highlighting in doc blocks in VSCode.

[the remark this answers is not in the document]
Comment (BendingBender) [reply index.d.ts]: Should I drop the block-level indentation as well?

[the remark this answers is not in the document]
Comment (sindresorhus) [reply index.d.ts]: > This, however, disables markdown syntax highlighting in doc blocks in VSCode.

I'm ok with that. The markdown is still correctly rendered in the preview for consumers.

https://microsoft.github.io/tsdoc/ correctly syntax highlights Markdown, even without the `*`, so it's probably just an oversight in VS Code.

File: index.d.ts (hunk 1 above)
Comment (BendingBender) [reply index.d.ts]: Sounds reasonable. Will do.

File: index.d.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (sindresorhus) [reply index.d.ts]: > Should I drop the block-level indentation as well?

The indentation should be the same as the code below.

**Comment by BendingBender** [comment]
Waiting for https://github.com/sindresorhus/type-fest/pull/5.

**Comment by sindresorhus** [comment]
> Waiting for sindresorhus/type-fest#5.

Published :)

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+	@example
+	```
+flags: {
+	unicorn: {
+		type: 'string',
+		alias: 'u',
+		default: 'rainbow'
+	}
+}
```
Comment (sindresorhus) [inline index.d.ts]: Sorry for being unclear, this should be indented the same as the backticks.

```
	```
	flags: {
		unicorn: {
			type: 'string',
			alias: 'u',
			default: 'rainbow'
		}
	}
```

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+	_You most likely don't need this option._
+	 */
+	readonly pkg?: {[key: string]: unknown};
+
+	/**
+	Custom arguments object.
+
+	@default process.argv.slice(2)
+	 */
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
	*/
```

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+//		r: true,
+//		unicorn: false,
+//		u: false,
+//		sparkles: true
+//	},
+//	…
+//}
+	 ```
+	 */
+	readonly booleanDefault?: boolean | null;
```
Comment (sindresorhus) [inline index.d.ts]: According to the docs this should support `undefined` too. It's unclear to me why `null` is mentioned in the docs though. Might be a mistake...

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+export interface Result {
+	/**
+	Non-flag arguments.
+	 */
+	input: string[];
+
+	/**
+	Flags converted to camelCase excluding aliases.
+	 */
+	flags: {[name: string]: unknown};
```
Comment (sindresorhus) [inline index.d.ts]: What your opinion on `{[name: string]: unknown};` vs `Record<string, unknown>`?

File: index.d.ts [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+	The help text used with `--help`.
+	 */
+	help: string;
+
+	/**
+	Show the help text and exit with code.
+
+	@param code The exit code to use. Default: `2`.
+	 */
+	showHelp(code?: number): void;
```
Comment (sindresorhus) [inline index.d.ts]: `code` => `exitCode` would be clear (I know you just copied from the readme)

File: index.d.ts (hunk 5 above)
Comment (BendingBender) [reply index.d.ts]: It's already optional, so `undefined` is already included.

File: index.d.ts (hunk 6 above)
Comment (BendingBender) [reply index.d.ts]: I think `{[name: string]: unknown};` is more common, so I'd prefer this.

File: index.d.ts (hunk 5 above)
Comment (sindresorhus) [reply index.d.ts]: Ah right... My mistake.

File: index.d.ts (hunk 6 above)
Comment (BendingBender) [reply index.d.ts]: And it allows naming the key.

File: index.d.ts (hunk 5 above)
Comment (BendingBender) [reply index.d.ts]: Included it for clarity.

**Comment by sindresorhus** [comment]
This looks great! :)

**Comment by sindresorhus** [comment]
Just FYI. It's probably going to take some time until I can publish this one as it's blocked by doing the next major release.

**Pull request merged**